crates/matrix-sdk-base/src/latest_event.rs

```diff
@@ -0,0 +1,197 @@
+                    EncryptedEventScheme, OlmV1Curve25519AesSha2Content, RoomEncryptedEventContent,
+                    SyncRoomEncryptedEvent,
+                },
+                message::{
+                    ImageMessageEventContent, MessageType, RoomMessageEventContent,
+                    SyncRoomMessageEvent,
+                },
+                topic::{RoomTopicEventContent, SyncRoomTopicEvent},
+                ImageInfo, MediaSource,
+            },
+            sticker::{StickerEventContent, SyncStickerEvent},
+            AnySyncMessageLikeEvent, AnySyncStateEvent, AnySyncTimelineEvent, EmptyStateKey,
+            MessageLikeUnsigned, OriginalSyncMessageLikeEvent, OriginalSyncStateEvent,
+            StateUnsigned,
+        },
+        owned_event_id, owned_mxc_uri, owned_user_id, MilliSecondsSinceUnixEpoch, UInt,
+    };
+
+    use crate::latest_event::{is_suitable_for_latest_event, PossibleLatestEvent};
+
+    #[test]
+    fn room_messages_are_suitable() {
+        let event = AnySyncTimelineEvent::MessageLike(AnySyncMessageLikeEvent::RoomMessage(
+            SyncRoomMessageEvent::Original(OriginalSyncMessageLikeEvent {
+                content: RoomMessageEventContent::new(MessageType::Image(
+                    ImageMessageEventContent::new(
+                        "".to_owned(),
+                        MediaSource::Plain(owned_mxc_uri!("mxc://example.com/1")),
+                    ),
+                )),
+                event_id: owned_event_id!("$1"),
+                sender: owned_user_id!("@a:b.c"),
+                origin_server_ts: MilliSecondsSinceUnixEpoch(UInt::new(2123).unwrap()),
+                unsigned: MessageLikeUnsigned::new(),
+            }),
+        ));
+        let m = assert_matches::assert_matches!(
+            is_suitable_for_latest_event(&event),
+            PossibleLatestEvent::YesMessageLike(m) => m
+        );
+
+        assert_eq!(m.content.msgtype.msgtype(), "m.image");
+    }
+
+    #[test]
+    fn different_types_of_messagelike_are_unsuitable() {
+        let event = AnySyncTimelineEvent::MessageLike(AnySyncMessageLikeEvent::Sticker(
+            SyncStickerEvent::Original(OriginalSyncMessageLikeEvent {
+                content: StickerEventContent::new(
+                    "sticker!".to_owned(),
+                    ImageInfo::new(),
+                    owned_mxc_uri!("mxc://example.com/1"),
+                ),
+                event_id: owned_event_id!("$1"),
+                sender: owned_user_id!("@a:b.c"),
+                origin_server_ts: MilliSecondsSinceUnixEpoch(UInt::new(2123).unwrap()),
+                unsigned: MessageLikeUnsigned::new(),
+            }),
+        ));
+
+        assert_matches!(
+            is_suitable_for_latest_event(&event),
+            PossibleLatestEvent::NoUnsupportedMessageLikeType
+        );
+    }
+
+    // TODO: I can't write this test because I can't construct a
+    // UnsignedRoomRedactionEvent.       I asked a question in the Ruma room.
```

Comment:
Docs state

> It is intentionally not possible to create an instance of this type other than through `Clone` or `Deserialize`.

And while I don't remember the original reason for this, can you just use `serde_json::from_value(json!({ ... })).unwrap()`?

Reply:
Thanks! Fixed in ef3b647d74

Reply:
BTW, this seems testing-hostile... if you did remember why it was done I'd be interested.